module/apmawssdkgo/sns.go

```diff
@@ -74,21 +74,44 @@ func (s *apmSNS) setAdditional(span *apm.Span) {
 }
 
 func getTopicName(req *request.Request) string {
-	// TODO: PhoneNumber is the third possibility, but I'm guessing we
-	// don't want to store that for customers?
-	arn := req.HTTPRequest.FormValue("TopicArn")
-	if arn == "" {
-		arn = req.HTTPRequest.FormValue("TargetArn")
+	// format: arn:aws:sns:us-east-2:123456789012:My-Topic
+	// should return My-Topic
+	if topicArn := req.HTTPRequest.FormValue("TopicArn"); topicArn != "" {
+		idx := strings.LastIndex(topicArn, ":")
+		if idx == -1 {
+			return ""
+		}
+
+		// special check for format: arn:aws:sns:us-east-2:123456789012/MyTopic
+		if slashIdx := strings.LastIndex(topicArn, "/"); slashIdx != -1 {
+			return topicArn[slashIdx+1:]
+		}
+
+		return topicArn[idx+1:]
 	}
 
-	// SNS ARN can be in the following formats:
-	// - arn:aws:sns:us-east-2:123456789012:MyTopic
-	// - arn:aws:sns:us-east-2:123456789012/MyTopic
-	parts := strings.Split(arn, "/")
-	if len(parts) == 1 {
-		parts = strings.Split(arn, ":")
+	// format: arn:aws:sns:us-west-2:123456789012:endpoint/GCM/gcmpushapp/5e3e9847-3183-3f18-a7e8-671c3a57d4b3
+	// should return endpoint/GCM/gcmpushapp
+	if targetArn := req.HTTPRequest.FormValue("TargetArn"); targetArn != "" {
```

Comment:
Maybe we should consider using `aws/arn.Parse`? https://github.com/aws/aws-sdk-go/blob/76296e15c619208361b3978b2337f5872f1ce01e/aws/arn/arn.go#L61

Reply:
I don't think we can replace the current method: `arn.Parse` seems to retrieve the resource which would be either `resource-type:resource-id` or `resource-type/resource-id`. We would still need to slice the string and lookup `:` and `/` since we're interested in the `resource-id`.

I'll merge this as is and maybe we can discuss improvements in a followup issue.